Internal error when groupping an empty frame

```
>>> import datatable as dt
>>> from datatable import f, by, count
>>> 
>>> DT = dt.Frame([[]])
>>> DT[:, count(f[0]), by(0)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Assertion 'max < rii->length || max == RowIndex::NA' failed in c/rowindex_array.cc, line 365
```

--- a/c/expr/by_node.cc
+++ b/c/expr/by_node.cc
@@ -153,11 +153,18 @@ void by_node::create_columns(EvalContext& ctx) {
 
 
 void by_node::execute(EvalContext& ctx) const {
-  if (cols.empty()) return;
+  if (cols.empty()) {
+    ctx.gb = Groupby::single_group(ctx.nrows());
+    return;
+  }
   const DataTable* dt0 = ctx.get_datatable(0);
   const RowIndex& ri0 = ctx.get_rowindex(0);
   if (ri0) {
     throw NotImplError() << "Groupby/sort cannot be combined with i expression";
+  }
+  if (dt0->nrows() == 0) {
+    // When grouping a Frame with 0 rows, produce a no-groups Groupby
+    return;
   }
   std::vector<sort_spec> spec;
   spec.reserve(cols.size());

--- a/c/expr/eval_context.cc
+++ b/c/expr/eval_context.cc
@@ -1,5 +1,5 @@
 //------------------------------------------------------------------------------
-// Copyright 2018 H2O.ai
+// Copyright 2018-2019 H2O.ai
 //
 // Permission is hereby granted, free of charge, to any person obtaining a
 // copy of this software and associated documentation files (the "Software"),
@@ -434,7 +434,7 @@ void EvalContext::apply_rowindex(const RowIndex& ri) {
 
 
 void EvalContext::apply_groupby(const Groupby& gb_) {
-  xassert(static_cast<size_t>(gb_.offsets_r()[gb_.ngroups()]) == nrows());
+  xassert(gb_.last_offset() == nrows());
   gb = gb_;
 }
 

--- a/c/expr/expr_reduce.cc
+++ b/c/expr/expr_reduce.cc
@@ -310,10 +310,7 @@ Column expr_reduce1::evaluate(EvalContext& ctx)
 {
   auto input_col = arg->evaluate(ctx);
   Groupby gb = ctx.get_groupby();
-  if (!gb) gb = Groupby::single_group(input_col.nrows());
-
   size_t out_nrows = gb.ngroups();
-  if (!out_nrows) out_nrows = 1;  // only when input_col has 0 rows
 
   if (opcode == Op::FIRST) {
     return reduce_first(input_col, gb);
@@ -374,10 +371,10 @@ Column expr_reduce0::evaluate(EvalContext& ctx) {
       const Groupby& grpby = ctx.get_groupby();
       size_t ng = grpby.ngroups();
       const int32_t* offsets = grpby.offsets_r();
-      res = Column::new_data_column(ng, SType::INT32);
-      auto d_res = static_cast<int32_t*>(res.get_data_editable());
+      res = Column::new_data_column(ng, SType::INT64);
+      auto d_res = static_cast<int64_t*>(res.get_data_editable());
       for (size_t i = 0; i < ng; ++i) {
-        d_res[i] = offsets[i + 1] - offsets[i];
+        d_res[i] = static_cast<int64_t>(offsets[i + 1] - offsets[i]);
       }
     } else {
       res = Column::new_data_column(1, SType::INT64);

--- a/c/groupby.cc
+++ b/c/groupby.cc
@@ -47,6 +47,11 @@ size_t Groupby::size() const noexcept {
   return n;
 }
 
+size_t Groupby::last_offset() const noexcept {
+  auto offs = offsets_r();
+  return offs? static_cast<size_t>(offs[n]) : 0;
+}
+
 Groupby::operator bool() const {
   return n != 0;
 }
@@ -60,6 +65,7 @@ void Groupby::get_group(size_t i, size_t* i0, size_t* i1) const {
 
 RowIndex Groupby::ungroup_rowindex() {
   const int32_t* offs = offsets_r();
+  if (!offs) return RowIndex();
   int32_t nrows = offs[n];
   arr32_t indices(static_cast<size_t>(nrows));
   int32_t* data = indices.data();

--- a/c/groupby.h
+++ b/c/groupby.h
@@ -28,6 +28,7 @@ class Groupby {
     const int32_t* offsets_r() const;
     size_t ngroups() const;
     size_t size() const noexcept;  // same as ngroups()
+    size_t last_offset() const noexcept;
     explicit operator bool() const;
     void get_group(size_t i, size_t* i0, size_t* i1) const;
 

--- a/docs/changelog/v-0-10-0.rst
+++ b/docs/changelog/v-0-10-0.rst
@@ -85,6 +85,16 @@ Frame
   incorrectly when some columns in the frame contained NAs, while others
   did not (#2050).
 
+- |fix| Groupby operation on an empty (0-rows) frame now works correctly,
+  returning a 0-row result frame (#2078). For example:
+
+    DT = dt.Frame(Id=[], Value=[])  # create a 0x2 frame
+    DT[:, sum(f.Value), by(f.Id)]   # produces a 0x2 frame
+    DT[:, sum(f.Value)]             # produces a 1x1 frame
+
+- |fix| The :meth:`count` reducer now always produces a column with stype
+  ``int64``. Before, it sometimes produced an ``int32`` column, and sometimes
+  an ``int64`` column.
 
 
 General

--- a/setup.cfg
+++ b/setup.cfg
@@ -11,4 +11,4 @@ inline-quotes = "
 [tool:pytest]
 minversion = 3.0
 norecursedirs = benchmarks build c ci docs microbench temp .benchmarks .cache .git .idea .venv
-python_files = test_*.py bench_*.py
+python_files = test_*.py test-*.py

--- a/tests/test-groups.py
+++ b/tests/test-groups.py
@@ -117,9 +117,17 @@ def test_group_empty_frame1():
 def test_group_empty_frame2():
     DT = dt.Frame(A=[])
     D1 = DT[:, count(), by(f.A)]
-    D2 = DT[:, count(f.A), by(f.A)]
     frame_integrity_check(D1)
+    assert D1.shape == (1, 0)
+    assert D1.stype == dt.int64
+
+
+def test_group_empty_frame3():
+    DT = dt.Frame(A=[])
+    D2 = DT[:, count(f.A), by(f.A)]
     frame_integrity_check(D2)
+    assert D2.shape == (1, 0)
+    assert D2.stype == dt.int64
 
 
 def test_groups_small1():

--- a/tests/test-groups.py
+++ b/tests/test-groups.py
@@ -118,16 +118,16 @@ def test_group_empty_frame2():
     DT = dt.Frame(A=[])
     D1 = DT[:, count(), by(f.A)]
     frame_integrity_check(D1)
-    assert D1.shape == (1, 0)
-    assert D1.stype == dt.int64
+    assert D1.shape == (0, 2)
+    assert D1.stypes == (DT.stype, dt.int64)
 
 
 def test_group_empty_frame3():
-    DT = dt.Frame(A=[])
+    DT = dt.Frame(A=[], stype=dt.float32)
     D2 = DT[:, count(f.A), by(f.A)]
     frame_integrity_check(D2)
-    assert D2.shape == (1, 0)
-    assert D2.stype == dt.int64
+    assert D2.shape == (0, 2)
+    assert D2.stypes == (DT.stype, dt.int64)
 
 
 def test_groups_small1():
@@ -322,6 +322,6 @@ def test_groupby_with_sort():
     R2 = DT[:, count(), by(f.A, f.B), sort(f.C)]
     R0 = dt.Frame(A=[1, 1, 2, 2, 3, 3],
                   B=[1, 2, 1, 2, 1, 2],
-                  C0=[2] * 6, stypes={"C0": dt.int32})
+                  C0=[2] * 6, stypes={"C0": dt.int64})
     assert_equals(R1, R0)
     assert_equals(R2, R0)

--- a/tests/test-groups.py
+++ b/tests/test-groups.py
@@ -1,7 +1,7 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
 #-------------------------------------------------------------------------------
-# Copyright 2018 H2O.ai
+# Copyright 2018-2019 H2O.ai
 #
 # Permission is hereby granted, free of charge, to any person obtaining a
 # copy of this software and associated documentation files (the "Software"),
@@ -30,81 +30,76 @@ from tests import same_iterables, assert_equals, isview
 
 
 
-def test_groups_internal2():
-    d0 = dt.Frame([[1,   5,   3,   2,   1,    3,   1,   1,   None],
+def test_groups1():
+    DT = dt.Frame([[1,   5,   3,   2,   1,    3,   1,   1,   None],
                    ["a", "b", "c", "a", None, "f", "b", "h", "d"]],
                   names=["A", "B"])
-    d1 = d0[:, :, by("A")]
-    # gb = d1.internal.groupby
-    # assert gb.ngroups == 5
-    # assert gb.group_sizes == [1, 4, 1, 2, 1]
+    d1 = DT[:, :, by("A")]
     assert_equals(
         d1, dt.Frame(A=[None, 1, 1, 1, 1, 2, 3, 3, 5],
                      B=["d", "a", None, "b", "h", "a", "c", "f", "b"]))
-    d2 = d0[:, :, by("B")]
-    # gb = d2.internal.groupby
-    # assert gb.ngroups == 7
-    # assert gb.group_sizes == [1, 2, 2, 1, 1, 1, 1]
+
+    d2 = DT[:, :, by("B")]
     assert_equals(
         d2, dt.Frame(B=[None, "a", "a", "b", "b", "c", "d", "f", "h"],
                      A=[1, 1, 2, 5, 1, 3, None, 3, 1]))
 
 
-def test_groups_internal3():
+def test_groups2():
     f0 = dt.Frame(A=[1, 2, 1, 3, 2, 2, 2, 1, 3, 1], B=range(10))
     f1 = f0[:, [f.B, f.A + f.B], by(f.A)]
     frame_integrity_check(f1)
     assert f1.to_list() == [[1, 1, 1, 1, 2, 2, 2, 2, 3, 3],
                             [0, 2, 7, 9, 1, 4, 5, 6, 3, 8],
                             [1, 3, 8, 10, 3, 6, 7, 8, 6, 11]]
-    # gb = f1.internal.groupby
-    # assert gb
-    # assert gb.ngroups == 3
-    # assert gb.group_sizes == [4, 4, 2]
-    # assert gb.group_offsets == [0, 4, 8, 10]
 
 
 @pytest.mark.parametrize("seed", [random.getrandbits(32)])
-def test_groups_internal4(seed):
+def test_groups3(seed):
     random.seed(seed)
     n = 100000
     src = [random.getrandbits(10) for _ in range(n)]
-    f0 = dt.Frame({"A": src})
+    f0 = dt.Frame(A=src, B=range(n))
     f1 = f0[:, :, by("A")]
     frame_integrity_check(f1)
-    # gb = f1.internal.groupby
-    # assert gb
-    # grp_offsets = gb.group_offsets
-    # ssrc = sorted(src)
-    # x_prev = -1
-    # for i in range(gb.ngroups):
-    #     s = set(ssrc[grp_offsets[i]:grp_offsets[i + 1]])
-    #     assert len(s) == 1
-    #     x = list(s)[0]
-    #     assert x != x_prev
-    #     x_prev = x
+    assert f1.shape == (n, 2)
+    assert f1.names == ("A", "B")
+    f1A, f1B = f1.to_list()
+    f1A.append(None)
+    curr_value = f1A[0]
+    group_start = 0
+    for i in range(n + 1):
+        if f1A[i] != curr_value:
+            group = f1B[group_start:i]
+            assert group == sorted(group)
+            curr_value = f1A[i]
+            group_start = i
+    assert group_start == n
+    assert curr_value is None
 
 
 @pytest.mark.parametrize("seed", [random.getrandbits(32)])
-def test_groups_internal5_strs(seed):
+def test_groups4(seed):
     random.seed(seed)
     n = 1000
     src = ["%x" % random.getrandbits(8) for _ in range(n)]
-    f0 = dt.Frame({"A": src})
+    f0 = dt.Frame(A=src, B=range(n))
     f1 = f0[:, :, by("A")]
     frame_integrity_check(f1)
-    # gb = f1.internal.groupby
-    # assert gb
-    # grp_offsets = gb.group_offsets
-    # ssrc = sorted(src)
-    # assert f1.to_list() == [ssrc]
-    # x_prev = None
-    # for i in range(gb.ngroups):
-    #     s = set(ssrc[grp_offsets[i]:grp_offsets[i + 1]])
-    #     assert len(s) == 1
-    #     x = list(s)[0]
-    #     assert x != x_prev
-    #     x_prev = x
+    assert f1.shape == (n, 2)
+    assert f1.names == ("A", "B")
+    f1A, f1B = f1.to_list()
+    f1A.append(None)
+    curr_value = f1A[0]
+    group_start = 0
+    for i in range(n + 1):
+        if f1A[i] != curr_value:
+            group = f1B[group_start:i]
+            assert group == sorted(group)
+            curr_value = f1A[i]
+            group_start = i
+    assert group_start == n
+    assert curr_value is None
 
 
 
@@ -112,7 +107,22 @@ def test_groups_internal5_strs(seed):
 # Groupby on small datasets
 #-------------------------------------------------------------------------------
 
-def test_groups1():
+def test_group_empty_frame1():
+    DT = dt.Frame(A=[])
+    assert DT.shape == (0, 1)
+    D1 = DT[:, :, by(f.A)]
+    frame_integrity_check(D1)
+
+
+def test_group_empty_frame2():
+    DT = dt.Frame(A=[])
+    D1 = DT[:, count(), by(f.A)]
+    D2 = DT[:, count(f.A), by(f.A)]
+    frame_integrity_check(D1)
+    frame_integrity_check(D2)
+
+
+def test_groups_small1():
     f0 = dt.Frame({"A": [1, 2, 1, 2, 1, 3, 1, 1],
                    "B": [0, 1, 2, 3, 4, 5, 6, 7]})
     f1 = f0[:, mean(f.B), by(f.A)]
@@ -141,14 +151,14 @@ def test_groups_autoexpand():
                             [2, 7, 0, 5, 13]]
 
 
-@pytest.mark.skip(reason="Issue #1586")
+@pytest.mark.xfail(reason="Issue #1586")
 def test_groupby_with_filter1():
-    f0 = dt.Frame({"KEY": [1, 2, 1, 2, 1, 2], "X": [-10, 2, 3, 0, 1, -7]})
+    f0 = dt.Frame(KEY=[1, 2, 1, 2, 1, 2], X=[-10, 2, 3, 0, 1, -7])
     f1 = f0[f.X > 0, sum(f.X), f.KEY]
     assert f1.to_list() == [[1, 2], [4, 2]]
 
 
-@pytest.mark.skip(reason="Issue #1586")
+@pytest.mark.xfail(reason="Issue #1586")
 def test_groupby_with_filter2():
     # Check that rowindex works even when applied to a view
     n = 10000

--- a/tests/test-groups.py
+++ b/tests/test-groups.py
@@ -130,6 +130,14 @@ def test_group_empty_frame3():
     assert D2.stypes == (DT.stype, dt.int64)
 
 
+def test_group_empty_frame4():
+    DT = dt.Frame(A=[], stype=dt.float32)
+    D2 = DT[:, sum(f.A), by(f.A)]
+    frame_integrity_check(D2)
+    assert D2.shape == (0, 2)
+    assert D2.stypes == (dt.float32, dt.float32)
+
+
 def test_groups_small1():
     f0 = dt.Frame({"A": [1, 2, 1, 2, 1, 3, 1, 1],
                    "B": [0, 1, 2, 3, 4, 5, 6, 7]})

